crates/ra_hir_expand/src/util.rs

```diff
@@ -0,0 +1,124 @@
+            QuoteError::IllegalChar => "illegal character",
+            QuoteError::UnrecognizedEscape => "unrecognized escape sequence",
+            QuoteError::InvalidUnicode => "invalid unicode code point",
+        }
+    }
+}
+
+/// Unquotes `s`.
+pub(crate) fn unquote(s: &str) -> Result<String, QuoteError> {
+    if s.chars().count() < 2 {
+        return Err(QuoteError::NotEnoughChars);
+    }
+
+    let quote = s.chars().next().unwrap();
+
+    if quote != '"' && quote != '\'' && quote != '`' {
+        return Err(QuoteError::UnrecognizedQuote);
+    }
+
+    if s.chars().last().unwrap() != quote {
+        return Err(QuoteError::UnexpectedEOF);
+    }
+
+    // removes quote characters
+    // the sanity checks performed above ensure that the quotes will be ASCII and this will not
+    // panic
+    let s = &s[1..s.len() - 1];
+
+    unescape(s, Some(quote))
+}
+
+/// Returns `s` after processing escapes such as `\n` and `\x00`.
+pub(crate) fn unescape(s: &str, illegal: Option<char>) -> Result<String, QuoteError> {
+    let mut chars = s.chars();
+    let mut unescaped = String::new();
+    loop {
+        match chars.next() {
+            None => break,
+            Some(c) => unescaped.push(match c {
+                _ if Some(c) == illegal => return Err(QuoteError::IllegalChar),
+                '\\' => match chars.next() {
+                    None => return Err(QuoteError::UnexpectedEOF),
+                    Some(c) => match c {
+                        _ if c == '\\' || c == '"' || c == '\'' || c == '`' => c,
+                        'a' => '\x07',
+                        'b' => '\x08',
+                        'f' => '\x0c',
+                        'n' => '\n',
+                        'r' => '\r',
+                        't' => '\t',
+                        'v' => '\x0b',
+                        // octal
+                        '0'..='9' => {
+                            let octal = c.to_string() + &take(&mut chars, 2);
+                            u8::from_str_radix(&octal, 8)
+                                .map_err(|_| QuoteError::UnrecognizedEscape)?
+                                as char
+                        }
+                        // hex
+                        'x' => {
+                            let hex = take(&mut chars, 2);
+                            u8::from_str_radix(&hex, 16)
+                                .map_err(|_| QuoteError::UnrecognizedEscape)?
+                                as char
+                        }
+                        // unicode
+                        'u' => decode_unicode(&take(&mut chars, 4))?,
+                        'U' => decode_unicode(&take(&mut chars, 8))?,
+                        _ => return Err(QuoteError::UnrecognizedEscape),
+                    },
+                },
+                _ => c,
+            }),
+        }
+    }
+
+    Ok(unescaped)
+}
+
+#[inline]
+// Iterator#take cannot be used because it consumes the iterator
```

Comment:
Just a quick note, but `iter.by_ref().take(n)` should do the trick

Reply:
Oh, i do not know this trick, thanks !